apiserver/facades/agent/uniter/uniter.go

```diff
@@ -2634,16 +2632,21 @@ func (u *UniterAPI) goalStateRelations(appName string, allRelations []*state.Rel
 	return result, nil
 }
 
-// goalStateUnits loops through all application units, and stores the
-// application GoalStateStatus in UnitsGoalState.
-func (u *UniterAPI) goalStateUnits(application *state.Application) (params.UnitsGoalState, error) {
+// goalStateUnits loops through all application units related to principalName,
+// and stores the goal state status in UnitsGoalState.
+func (u *UniterAPI) goalStateUnits(app *state.Application, principalName string) (params.UnitsGoalState, error) {
 
-	allUnits, err := application.AllUnits()
+	allUnits, err := app.AllUnits()
 	if err != nil {
 		return nil, err
 	}
 	unitsGoalState := params.UnitsGoalState{}
 	for _, unit := range allUnits {
+		// Ignore subordinates belonging to other units.
+		pn, ok := unit.PrincipalName()
+		if ok && pn != principalName {
+			continue
+		}
```

Comment:
Presumably this is the core of the change, and the rest is just being able to pass in principalName into this code, and a small amount of refactoring.

Reply:
yup